compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/parser/BLangNodeTransformer.java

```diff
@@ -2297,6 +2304,99 @@ public BLangNode transform(ArrayTypeDescriptorNode arrayTypeDescriptorNode) {
         return arrayTypeNode;
     }
 
+    public void addEnumDeclarationNode(
+            EnumDeclarationNode enumDeclarationNode,
+            BLangCompilationUnit compInit) {
+
+        Boolean publicQualifier = false;
+        if (enumDeclarationNode.qualifier() != null &&
+                enumDeclarationNode.qualifier().kind()
+                        == SyntaxKind.PUBLIC_KEYWORD) {
+            publicQualifier = true;
+        }
+        for (Node member : enumDeclarationNode.enumMemberList()) {
+            compInit.addTopLevelNode(transformEnumMember(
+                    (EnumMemberNode) member,
+                    publicQualifier));
+        }
+
+        BLangTypeDefinition bLangTypeDefinition =
+                (BLangTypeDefinition) TreeBuilder.createTypeDefinition();
+        if (publicQualifier) {
+            bLangTypeDefinition.flagSet.add(Flag.PUBLIC);
+        }
+
+        bLangTypeDefinition.setName(
+                (BLangIdentifier)transform(enumDeclarationNode.identifier()));
+
+        BLangUnionTypeNode bLangUnionTypeNode =
+                (BLangUnionTypeNode) TreeBuilder.createUnionTypeNode();
+        for (Node member : enumDeclarationNode.enumMemberList()) {
+            bLangUnionTypeNode.memberTypeNodes.
+                    add(createTypeNode(
+                            ((EnumMemberNode) member).identifier()));
+        }
+        Collections.reverse(bLangUnionTypeNode.memberTypeNodes);
+        bLangTypeDefinition.setTypeNode(bLangUnionTypeNode);
+        compInit.addTopLevelNode(bLangTypeDefinition);
+    }
+
+    public BLangConstant transformEnumMember(EnumMemberNode member,
+                                             Boolean publicQualifier) {
+        BLangConstant bLangConstant =
+                (BLangConstant) TreeBuilder.createConstantNode();
+
+        bLangConstant.flagSet.add(Flag.CONSTANT);
+
+        if (publicQualifier) {
+            bLangConstant.flagSet.add(Flag.PUBLIC);
+        }
+        BLangTypeDefinition assosiatedTypeDef =
```

Comment:
see if you can reuse the `deSugarTypeAsUserDefType` function instead.

Reply:
sure :)